### ISSUE TITLE
Add  certbot command line options

### ISSUE THREAD

**Issue by keckelt** [phovea/phovea_landing_page#10]
Certbot Docu: https://certbot.eff.org/docs/using.html#certbot-command-line-options

I would add `--non-interactive`:

> run without ever asking for user input. This may require additional command line flags; the client will try to explain which ones are required if it finds one missing (default: False)

certbot itself suggests this:
```
nginx_landing_page_1  | Running initial certificate request... 
nginx_landing_page_1  | certbot certonly -d tourguide.caleydoapp.org      --standalone --text     --email bot@caleydo.org --agree-tos     --expand 
nginx_landing_page_1  | Saving debug log to /var/log/letsencrypt/letsencrypt.log
nginx_landing_page_1  | Plugins selected: Authenticator standalone, Installer None
nginx_landing_page_1  | Skipped user interaction because Certbot doesn't appear to be running in a terminal. You should probably include --non-interactive or --force-interactive on the command line.
```

and `--keep-until-expiring` or `--force-renewal` to avoid errors due to input prompt in docker:
> If the requested certificate matches an existing certificate, always keep the existing one until it is due for renewal (for the 'run' subcommand this means reinstall the existing certificate). (default: Ask)

This was the case when I re-requested a single certificate after a few minutes.

**Comment by thinkh** [comment]
Sounds good. Could you please add this and file a PR?

**Comment by keckelt** [comment]
The renewal of certificates via cron is slightly different than the initial certificate request:

Certificates are currently renewed with [certbot renew](https://certbot.eff.org/docs/using.html#renewing-certificates).

> renew is suitable (and designed) for automated use, to allow your system to automatically renew each certificate when appropriate. Since renew only renews certificates that are near expiry it can be run as frequently as you want - since it will usually take no action.

When appropriate means:
> This command attempts to renew any previously-obtained certificates that expire in less than 30 days.

I.e. nothing has to be done here.

**Comment by keckelt** [comment]
I tried renewing the certificates from the container command line, using the command from [entry_point.sh](https://github.com/phovea/phovea_landing_page/blob/master/entry_point/entry_point.sh#L106):
```
certbot renew --webroot --webroot-path /etc/letsencrypt/webrootauth/ --post-hook "/usr/sbin/nginx -s reload"
```
and also simply with `certbot renew --force-renewal`, both failed with:
```
Attempting to renew cert (caleydoapp.org) from /etc/letsencrypt/renewal/caleydoapp.org.conf produced an unexpected error: Problem binding to port 80: Could not bind to IPv4 or IPv6.. Skipping.
All renewal attempts failed. The following certs could not be renewed:
  /etc/letsencrypt/live/caleydoapp.org/fullchain.pem (failure)
```

**Comment by keckelt** [comment]
The currently used certbot version on AWS is: 
```sh
/ # certbot --version
certbot 0.30.2
```
whereas the [current version](https://github.com/certbot/certbot/releases) 0.39.
because there the image is still based on Alpine 3.9, see [the distributions package list](https://pkgs.alpinelinux.org/packages?name=certbot&branch=v3.9).


This might be critical because let's encrypt had an API change and certbot is failing exactly there, see: https://community.letsencrypt.org/t/malformed-account-id-in-keyid-header-url-https-acme-v02-api-letsencrypt-org-acme-acct/99790 and https://community.letsencrypt.org/t/acme-v2-scheduled-removal-of-id-from-account-objects/94744

Accessing the [failing API endpoint](https://acme-v02.api.letsencrypt.org/acme/acct/64439940) myself i also get an 405 error: `Method not allowed`.

**Comment by keckelt** [comment]
Our Dockerfile is not based on a specific version of alpine: 
https://github.com/phovea/phovea_landing_page/blob/master/Dockerfile#L1

Therefore I updated the container that is running on our JKU server for tourguide.caleydoapp.org and tried it there (also less domains registered there) to [Alpine 3.10](https://alpinelinux.org/posts/Alpine-3.10.0-released.html) with [certbot v0.35.1](https://pkgs.alpinelinux.org/packages?name=certbot&branch=v3.10).

I test the renewal there now.

**Comment by keckelt** [comment]
nginx command line options: https://www.nginx.com/resources/wiki/start/topics/tutorials/commandline/

The server is started with `/usr/sbin/nginx -g 'daemon off;'`
The server reloads after certificate renewal with `/usr/sbin/nginx -s reload`, but that fails

`daemon off` keeps nginx in foreground --> docker container exits when nginx stops
>For normal production (on a server), use the default daemon on; directive so the Nginx server will start in the background. In this way Nginx and other services are running and talking to each other. One server runs many services.
For Docker containers (or for debugging), the daemon off; directive tells Nginx to stay in the foreground. For containers this is useful as best practice is for one container = one process. One server (container) has only one service.

**Comment by ghost** [comment]
hallo klaus; ich sehe erst jetzt das du an dem arbeitest; ich hab mich da auch schon etwas gespielt, wobei bei mir das renewal funktioniert hat (allerdings nicht der cronjob dazu :-) ); sollen wir uns das gemeinsam ansehen

**Comment by keckelt** [comment]
Gerne!

---

With the Alpine 3.10/certbot 0.35 container contained I had error messages for the post hook command that renews nginx's config:
Command
```
certbot renew --webroot --webroot-path /etc/letsencrypt/webrootauth/ --post-hook "/usr/sbin/nginx -s reload"
```

Error:
```
Saving debug log to /var/log/letsencrypt/letsencrypt.log

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Processing /etc/letsencrypt/renewal/tourguide.caleydoapp.org.conf
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Plugins selected: Authenticator webroot, Installer None
Renewing an existing certificate

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
new certificate deployed without reload, fullchain is
/etc/letsencrypt/live/tourguide.caleydoapp.org/fullchain.pem
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Congratulations, all renewals succeeded. The following certs have been renewed:
  /etc/letsencrypt/live/tourguide.caleydoapp.org/fullchain.pem (success)
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Running post-hook command: /usr/sbin/nginx -s reload
Error output from post-hook command nginx:
2019/10/16 12:30:04 [notice] 145#145: signal process started
```

But (!), it's just error ouput, not an error (i.e. nginx writes to stderr instead of stdout), the return code is still `0` and thus successful, as is the renewal!

You can check if the renwal worked with [crt.sh](https://crt.sh/) or `echo | openssl s_client -connect tourguide.caleydoapp.org:443 | openssl x509 -noout -dates`

Without reloading nginx via the post hook command the returned certificates were not up to date.

**Comment by keckelt** [comment]
Certbot's post-hook command is run even if the renewal was not succesful:
> When Certbot detects that a certificate is due for renewal, --pre-hook and --post-hook hooks run before and after each attempt to renew it. If you want your hook to run only after a successful renewal, use --deploy-hook.

Src: https://certbot.eff.org/docs/using.html#renewing-certificates

So a deploy hook might be better to avoid unnecessary reloading of the configuration.

**Comment by keckelt** [comment]
Also: certbot seems to store previously specified hooks.

If you execute `certbot rewnew --post-hook "echo hello world"` for example and later simply call `certbot renew` it will still execute the previous post-hook.
You can get rid of the old hooks by passing `--post-hook ""`

**Comment by keckelt** [comment]
Executing this one manually worked as expected inside the landing page container on the JKU Server: 
```bash
 certbot renew --force-renewal --webroot --webroot-path /etc/letsencrypt/webrootauth/ --deploy-hook "/usr/sbin/nginx -s reload"
```

**Comment by keckelt** [comment]
Alpine [usually](https://www.osetc.com/en/how-to-start-stop-restart-services-on-alpine-linux.html) uses `openrc` to start stop services (like crond):
https://wiki.alpinelinux.org/wiki/Alpine_Linux:FAQ#My_cron_jobs_don.27t_run.3F 

**But**: `openrc` and its commands (`rc-status`, `rc-service`, ...) [are not included in the docker image](https://github.com/gliderlabs/docker-alpine/issues/183).

Putting simple logging scripts into the /etc/periodic/<interval> directories works though, e.g.:
```sh
#!/bin/sh
echo "daily  $(date)" >> /var/log/crontest.log
```

And `ps aux`shows _crond_ is running:
```sh
/ # ps aux
PID   USER     TIME  COMMAND
    1 root      0:00 nginx: master process /usr/sbin/nginx -g daemon off;
   27 root      0:01 /usr/sbin/crond -f -d 8
...
````

**Comment by keckelt** [comment]
I'm not sure why the `--webroot-path` points to the let's encrypt directory ` /etc/letsencrypt/webrootauth/`.

Maybe because we use a reverse proxy to the apps and [a workaround ](https://community.letsencrypt.org/t/using-the-webroot-domain-verification-method/1445/7)is used?

On the jku server it works, but the same command fails on aws. Not getting the expected responses for the subdomains:
```
APPS:
cohort.caleydoapp.org
cohort-daily.caleydoapp.org
cohort-student.caleydoapp.org
cohort-student-daily.caleydoapp.org
confusionflow.caleydoapp.org
confusionflow-daily.caleydoapp.org
gapminder.caleydoapp.org
gapminder-retrieval.caleydoapp.org
lineage.caleydoapp.org
ordino-daily.caleydoapp.org
ordino-p2f.caleydoapp.org
ordino-p2f-sinpro.caleydoapp.org
ordino-retrieval.caleydoapp.org
ordino-student.caleydoapp.org
p2f-moldsonics.caleydoapp.org
pathfinder.caleydoapp.org
stratomex.caleydoapp.org
taco.caleydoapp.org
thermalplot.caleydoapp.org
tourdino.caleydoapp.org
vega-gapminder.caleydoapp.org

FORWARDS:
ordino-touring.caleydoapp.org
tourguide-ordino.caleydoapp.org
avocado.caleydo.org
cancer.caleydo.org
clue.caleydo.org
connecting-the-dots.caleydo.org
contour.caleydo.org
furby.caleydo.org
help.caleydo.org
hidden-content.caleydo.org
knowledge-pearls.caleydo.org
lineup.caleydo.org
ordino.caleydo.org
pathfinder.caleydo.org
pathways.caleydo.org
phovea.caleydo.org
projects.caleydo.org
stratomex.caleydo.org
taco.caleydo.org
tourdino.caleydo.org
tourguide.caleydo.org
visonfire.caleydo.org
```


Edit: check https://community.letsencrypt.org/t/could-not-connect-to-http-domain-name-org-well-known-acme-challenge-xxxx/15748

**Comment by keckelt** [comment]
> On the jku server it works, but the same command fails on aws. Not getting the expected responses for the subdomains:

Fixed! The image on the EC2 instance was not up to date!
I connected to the instance, removed the image and then started the task again and now refreshing the certificates manually from command line works.

I'll edit the command to force a rewnew when its executed by cron on the weekend and then let's see if something happens.

Command with forced renewal of certificates:
```sh
certbot renew --force-renewal --webroot --webroot-path /etc/letsencrypt/webrootauth/ --post-hook "/usr/sbin/nginx -s reload"
```

**Comment by keckelt** [comment]
The landing_page container on our aws ec2 instance is still based on Alpine 3.9.0 - don't know why that wasn't updated to 3.10.
So it is also still using certbot v0.30.2, but the command succeeds. So my assumption in [the comment above](https://github.com/phovea/phovea_landing_page/issues/10#issuecomment-542651980) (that the old version is causing problems) is wrong, i guess.

**Comment by keckelt** [comment]
I'm also removing `--text` as this no longer needed since [v.0.10.0](https://github.com/certbot/certbot/blob/a27b1137a55db1fdf09116e7849e464a2f9e0c28/CHANGELOG.md#0100---2017-01-11)
